plugin.py

```diff
@@ -33,7 +45,77 @@ def on_workspace_configuration(self, params, configuration) -> None:
-                        configuration['workspaceFolder'] = folder.to_lsp()
+                        workspace_folder = folder
                         break
+                if workspace_folder:
+                    configuration['workspaceFolder'] = workspace_folder.to_lsp()
+                self.resolve_working_directory(configuration, scope_uri, workspace_folder)
+
+    def resolve_working_directory(self, configuration, scope_uri, workspace_folder) -> None:
+        working_directories = configuration.get('workingDirectories', None)
+        if isinstance(working_directories, list):
+            working_directory = None
+            workspace_folder_path = workspace_folder.path if workspace_folder else None
+            for entry in working_directories:
+                directory = None
+                no_cwd = False
+                if isinstance(entry, str):
+                    directory = entry
+                elif self.is_directory_item(entry):
+                    directory = entry.directory
+                    if isinstance(entry.get('!cwd', None), bool):
+                        no_cwd = entry['!cwd']
+                elif self.is_pattern_item(entry):
+                    print('LSP-eslint: workingDirectories configuration that uses "pattern" is not supported')
+                    pass
+                elif self.is_mode_item(entry):
+                    working_directory = entry
+                    continue
+
+                directory_value = None
+                if directory:
+                    file_path = uri_to_filename(scope_uri)
+                    if directory:
+                        directory = self.to_os_path(directory)
+                        if not os.path.isabs(directory) and workspace_folder_path:
+                            # Trailing '' will add trailing slash if missing.
+                            directory = os.path.join(workspace_folder_path, directory, '')
+                        if file_path.startswith(directory):
+                            directory_value = directory
+
+                if directory_value:
+                    if not working_directory or self.is_mode_item(working_directory):
+                        working_directory = { 'directory': directory_value, '!cwd': no_cwd }
+                    else:
+                        if len(working_directory['directory']) < len(directory_value):
+                            working_directory['directory'] = directory_value
+                            working_directory['!cwd'] = no_cwd
+            configuration['workingDirectory'] = working_directory
+
+    def is_directory_item(self, item) -> bool:
+        if isinstance(item, dict):
+            directory = item.get('directory', None)
+            not_cwd = item.get('!cwd', None)
+            return isinstance(directory, str) and (isinstance(not_cwd, bool) or not_cwd == None)
+        return False
+
+    def is_pattern_item(self, item) -> bool:
```

Comment:
This function looks very similar to is_directory_item, can we turn it into something that doesn’t have a lot of copy-pasting?

Reply:
Done something about that.